recipe-client-addon/lib/Sampling.jsm

```diff
@@ -57,25 +55,72 @@ function bufferToHex(buffer) {
   return hexCodes.join("");
 }
 
+function rangeBucketSample(inputHash, minBucket, maxBucket, bucketCount) {
+  const minHash = fractionToKey(minBucket / bucketCount);
+  const maxHash = fractionToKey(maxBucket / bucketCount);
+  return (minHash <= inputHash) && (inputHash < maxHash);
+}
+
+/**
+ * Return a promise for a truncated hash of `data`.
+ */
+const truncatedHash = Task.async(function* (data) {
+  const hasher = crypto.subtle;
+  const input = new TextEncoder("utf-8").encode(JSON.stringify(data));
+  const hash = yield hasher.digest("SHA-256", input);
+    // truncate hash to 12 characters (2^48), because the full hash is larger
+    // than JS can meaningfully represent as a number.
+  return bufferToHex(hash).slice(0, 12);
+});
+
 this.Sampling = {
-  stableSample(input, rate) {
-    const hasher = crypto.subtle;
-
-    return hasher.digest("SHA-256", new TextEncoder("utf-8").encode(JSON.stringify(input)))
-      .then(hash => {
-        // truncate hash to 12 characters (2^48)
-        const inputHash = bufferToHex(hash).slice(0, 12);
-        const samplePoint = fractionToKey(rate);
-
-        if (samplePoint.length !== 12 || inputHash.length !== 12) {
-          throw new Error("Unexpected hash length");
-        }
-
-        return inputHash < samplePoint;
-
-      })
-      .catch(error => {
```

Comment:
I think logging when we hit an error in evaluating a filter expression is good, but I don't really care whether it's in the utility methods that we expose, or done once by the JEXL-evaluation code itself. Will we still do that somewhere once this is removed?